### PR TITLE
Add `name` and `geo` to User

### PR DESCRIPTION
**Pull request by denrase**
## :scroll: Description

- Add property `name` to `User`
- Add class `Geo`
- Add property `geo` to `User`

## :bulb: Motivation and Context

Relates to https://github.com/getsentry/sentry-dart/issues/1065

## :green_heart: How did you test it?

Unit tests

## :pencil: Checklist

You have to check all boxes before merging:

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [x] No breaking change or entry added to the changelog.
- [x] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

## :crystal_ball: Next steps

Set values correctly when syncing user from hybrid SDK

### CODE REVIEW

File: Sources/Sentry/SentryUser.m
Comment (philipphofmann) [inline Sources/Sentry/SentryUser.m:34]: `m`: We need to copy the geo cause the property `geo` uses `strong`.
```suggestion
            copy.geo = self.geo.copy;
```

File: Sources/Sentry/SentryGeo.m
Comment (philipphofmann) [inline Sources/Sentry/SentryGeo.m:36]: `l`: No need to follow the setValue forKey pattern. I like this syntax more, but it's the same. Up to you.
```suggestion
        serializedData[@"city"] = self.city;
        serializedData[@"country_code"] = self.countryCode;
        serializedData[@"region"] = self.region;
```

File: Sources/Sentry/SentryGeo.m
Comment (philipphofmann) [inline Sources/Sentry/SentryGeo.m:31]: `l`: A bit shorter if you want.
```suggestion
    NSMutableDictionary *serializedData = [NSMutableDictionary dictionary];
```

File: Tests/SentryTests/Protocol/SentryGeoTests.swift
Comment (philipphofmann) [inline Tests/SentryTests/Protocol/SentryGeoTests.swift:3]: Beautiful tests, how did you come up with these 😄🙃 ?

**Review by philipphofmann — COMMENTED** [review]
Almost LGTM @denrase.

File: Tests/SentryTests/Protocol/SentryGeoTests.swift
Comment (denrase) [reply Tests/SentryTests/Protocol/SentryGeoTests.swift:3]: You could say i've got a taste....


... for Copy & Paste 😎

File: Sources/Sentry/Public/SentryUser.h
Comment (denrase) [reply Sources/Sentry/Public/SentryUser.h:40]: We do have the `name` property and it was added in this PR: https://github.com/getsentry/sentry-dart/pull/1038 Maybe @marandaneto has more context?

**Review by philipphofmann — COMMENTED** [review]
This [comment](https://github.com/getsentry/sentry-cocoa/pull/2710#discussion_r1113261729) stops me from giving this an approval.

File: Sources/Sentry/Public/SentryUser.h
Comment (marandaneto) [reply Sources/Sentry/Public/SentryUser.h:40]: You can check https://develop.sentry.dev/sdk/event-payloads/types/#typedef-User
https://develop.sentry.dev/sdk/event-payloads/user/ is often out of date.
Going to `types/#typedef-...` is the living Relay code docs.

**Review by philipphofmann — APPROVED** [review]
LGTM

File: Sources/Sentry/SentryGeo.m
Comment (armcknight) [inline Sources/Sentry/SentryGeo.m:60]: I know this comment is copypasted from SentryUser, but I'm not sure what it describes is sufficient, since those properties could just as easily change in between those copy operations as they could during evaluation of the `if` statements:
1. Call `[a isEqualToGeo:b]`
2. copy `b.city` into `otherCity`
3. something modifies `b.countryCode` while comparing `a.city` to `otherCity`
4. copy `b.countryCode` into `otherCountryCode`

Now we're in an inconsistent state because `b` changed while it was being compared to `a`, and maybe has a city of London, England, but now has a country code of Japan. Is this a problem @philipphofmann?

I wonder if it doesn't make more sense to just compare their hashes after the self/nil checks, to implicitly compare their member data, since those are created in `@synchronized` blocks.

File: Sources/Sentry/Public/SentryGeo.h
Comment (armcknight) [inline Sources/Sentry/Public/SentryGeo.h:23]: Where would this come from? Could you provide some examples in the headerdoc?

File: Sources/Sentry/Public/SentryGeo.h
Comment (armcknight) [inline Sources/Sentry/Public/SentryGeo.h:18]: I wonder if we can't just grab the `NSLocale` object that would probably be used to populate this, so we can get any other info from that we need at a later date, as well as ensure the right value is actually set in our `serialize` implementation (there's a lot going on in NSLocale and some people might not get it right).

File: Sources/Sentry/Public/SentryUser.h
Comment (armcknight) [inline Sources/Sentry/Public/SentryUser.h:40]:
```suggestion
@property (nullable, atomic, copy) NSString *name;
```

**Review by armcknight — APPROVED** [review]
Some minor improvements and a question about the equality comparison, but otherwise LGTM.

File: Sources/Sentry/SentryGeo.m
Comment (denrase) [reply Sources/Sentry/SentryGeo.m:12]: The compiler complains when not providing this implementation.

Bildschirm­foto 2023-03-07 um 13 39 17

File: Sources/Sentry/Public/SentryGeo.h
Comment (denrase) [reply Sources/Sentry/Public/SentryGeo.h:23]: I can only point you to @marandaneto's answer here: https://github.com/getsentry/sentry-cocoa/pull/2710#discussion_r1118840132

Comment (marandaneto) [reply Sources/Sentry/Public/SentryGeo.h:23]: This is automatically filled during ingestion when the `ipAddress` is set or inferred by the server.
One can manually set their own `geo` data tho, this is about that.
For example, a user might be located at X but using a VPN at Y.
Sometimes the inferred data isn't the preferred one.

File: Sources/Sentry/SentryGeo.m
Comment (brustolin) [reply Sources/Sentry/SentryGeo.m:12]: If you don't have a reason to override `init` you can also remove it from the header, then you don't get this error.

File: Sources/Sentry/Public/SentryGeo.h
Comment (armcknight) [reply Sources/Sentry/Public/SentryGeo.h:23]: I guess what I'm really wondering is what is the expected format of data that would be in this field. Examples for countryCode would be `US`, `CZ` etc, I'm just not sure what would go in here. Is it just arbitrary and only meaningful to each separate app developer depending on their needs?

File: Sources/Sentry/SentryGeo.m
Comment (denrase) [reply Sources/Sentry/SentryGeo.m:55]: Is this an issue that is occurring with the other classes that follow this approach? If so we should probably look at this throughout the SDK, which would warrant a seperate PR. If there's no issue i'd rather leave the approach the same. WDYT?

File: Sources/Sentry/Public/SentryGeo.h
Comment (denrase) [reply Sources/Sentry/Public/SentryGeo.h:18]: As i understood this is filled out manually by users, so i'm not sure if we should do things automatically to populate these fields. What other info that we might need are you refering to?

File: Sources/Sentry/SentryGeo.m
Comment (armcknight) [reply Sources/Sentry/SentryGeo.m:55]: Yes, I agree it would be best in a separate issue/PR. I'll split out further discussion elsewhere.

File: Sources/Sentry/Public/SentryGeo.h
Comment (armcknight) [reply Sources/Sentry/Public/SentryGeo.h:18]: I suppose even if we know sure we won't want anything other than country code, we could still just get it directly from the NSLocale. I've seen plenty of mistakes mixing up country codes, language codes and locale identifiers. I would ask for some validation logic, but that would be more work than us just using the locale object in a known safe way.

> this is filled out manually by users

Do you mean end-users of apps consuming this SDK? Or users of the SDK, as in app developers? If the former, even more reason to validate input, but again, that's more work.

File: Sources/Sentry/Public/SentryGeo.h
Comment (marandaneto) [reply Sources/Sentry/Public/SentryGeo.h:23]: @armcknight it's in the spec that I linked to in my [comment above](https://github.com/getsentry/sentry-cocoa/pull/2710#discussion_r1127816190).
https://develop.sentry.dev/sdk/event-payloads/types/#geo

File: Sources/Sentry/Public/SentryGeo.h
Comment (marandaneto) [reply Sources/Sentry/Public/SentryGeo.h:18]: the server infers that automatically from the IP if provided, this should only be set if the user wanna override the inferred value by the server as explained here https://github.com/getsentry/sentry-cocoa/pull/2710#discussion_r1127830155

File: Sources/Sentry/SentryGeo.m
Comment (philipphofmann) [reply Sources/Sentry/SentryGeo.m:55]: I think we can remove the comment. I can't recall why I added it with https://github.com/getsentry/sentry-cocoa/pull/888. I think what you described in https://github.com/getsentry/sentry-cocoa/pull/2710#discussion_r1127523845 shouldn't be a problem. How likely is it that somebody modifies the user from multiple threads simultaneously and calls equals on them? I think we can remove all the `@synchronized` blocks in the user, but we should do this in another PR. @denrase, I think you should remove the `synchronized` for the SentryGeo class before merging this PR.

Comment (philipphofmann) [reply Sources/Sentry/SentryGeo.m:55]: I removed the synchronized from the user https://github.com/getsentry/sentry-cocoa/pull/2809.

File: Sources/Sentry/Public/SentryGeo.h
Comment (philipphofmann) [reply Sources/Sentry/Public/SentryGeo.h:23]: The spec lacks examples, @marandaneto. Could you please add some, as you added the spec https://github.com/getsentry/develop/pull/705?

**Review by philipphofmann — COMMENTED** [review]
@denrase, I think this PR should be ready for review. After you remove the `synchronize` blocks and maybe add examples for the region, I think you can merge this.

**Review by philipphofmann — COMMENTED** [review]
LGTM, thank you @denrase 😺

**Review by philipphofmann — APPROVED** [review]
Ups forgot to LGTM

File: Sources/Sentry/Public/SentryGeo.h
Comment (marandaneto) [reply Sources/Sentry/Public/SentryGeo.h:23]: Yep, can do it, but lets not block this PR, example:

```
{
  "geo": {
    "country_code": "US",
    "city": "Ashburn",
    "region": "San Francisco"
  }
}
```

Comment (philipphofmann) [reply Sources/Sentry/Public/SentryGeo.h:23]: @denrase, please add it to the code comments/docs in this repo. Thanks @marandaneto

**Pull request merged**
